Index: arp_hml/src/orocos/motor/Faulhaber3268Bx4.cpp
===================================================================
--- arp_hml/src/orocos/motor/Faulhaber3268Bx4.cpp	(revision 1301)
+++ arp_hml/src/orocos/motor/Faulhaber3268Bx4.cpp	(working copy)
@@ -186,6 +186,7 @@
 
     //publication de la vitesse
 	outVelocity.write( ArdMotorItf::getSpeedMeasure() );
+	outDEBUGSpeedCmd.write(m_speedCommand );
     //lecture du courant
     outTorque.write( ArdMotorItf::getTorqueMeasure() );
 
@@ -810,6 +811,8 @@
     addPort("outHomingDone",outHomingDone)
         .doc("Is true when the Homing sequence is finished. It has no sense when the motor is not in HOMING mode");
 
+    addPort("outDEBUGSpeedCmd",outDEBUGSpeedCmd);
+
     addOperation("ooEnableDrive", &Faulhaber3268Bx4::enableDrive,this, OwnThread )
         .doc("Activate the power on the motor. Without a speed command it acts as a brake. Returns false if the component is not running.");
     addOperation("ooDisableDrive", &Faulhaber3268Bx4::disableDrive,this, OwnThread )
Index: arp_hml/src/orocos/motor/Faulhaber3268Bx4.hpp
===================================================================
--- arp_hml/src/orocos/motor/Faulhaber3268Bx4.hpp	(revision 1301)
+++ arp_hml/src/orocos/motor/Faulhaber3268Bx4.hpp	(working copy)
@@ -122,6 +122,9 @@
         OutputPort<bool> outMaxTorqueTimeout;
         /** Is true when the Homing sequence is finished. It has no sense when the motor is not in HOMING mode */
         OutputPort<bool> outHomingDone;
+
+        OutputPort<double> outDEBUGSpeedCmd;
+
         /**
          * Limits the torque on the motor via a current limitation.
          * @param ampValue : the maximal current that the motor can require in A. Should be in ]0.2;10[
Index: arp_master/script/orocos/deployment/lua/telemetry.lua
===================================================================
--- arp_master/script/orocos/deployment/lua/telemetry.lua	(revision 1301)
+++ arp_master/script/orocos/deployment/lua/telemetry.lua	(working copy)
@@ -86,6 +86,7 @@
 	Reporting:reportPort("MotionControl","outDEBUGLinSpeedCorrection")
 	Reporting:reportPort("MotionControl","outDEBUGAngSpeedCorrection")
 	Reporting:reportPort("MotionControl","outDEBUGSaturation")
+	Reporting:reportPort("MotionControl","outDEBUGMode")
 end
 
 
@@ -101,8 +102,8 @@
 	print("====================")
 	print("début déploiment telemetry")
 	--Telemetry:reportJoystick()
-	--Telemetry:reportRobotCmd()
-	--Telemetry:reportRobotState()
+	Telemetry:reportRobotCmd()
+	Telemetry:reportRobotState()
 	--Telemetry:reportMotorState()
 	--Telemetry:reportTiming()
 	Telemetry:reportOmnidirect()
