Index: arp_rlu/src/components/ObstacleManager.cpp
===================================================================
--- arp_rlu/src/components/ObstacleManager.cpp	(revision 1407)
+++ arp_rlu/src/components/ObstacleManager.cpp	(working copy)
@@ -145,7 +145,16 @@
 //        }
 //        LOG( Info ) << ss.str() << endlog();
 //    }
-
+    std::vector<arp_math::EstimatedPose2D> dummy;
+    opponents = dummy;
+    arp_math::EstimatedPose2D oppDummy;
+    oppDummy.x(-0.500);
+    oppDummy.y(0.500);
+    opponents.push_back( oppDummy );
+    arp_math::EstimatedPose2D oppDummy2;
+    oppDummy2.x(0.500);
+    oppDummy2.y(0.500);
+    opponents.push_back( oppDummy2 );
     outOpponents.write(opponents);
 }
 
